src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs

```diff
@@ -57,6 +57,54 @@ internal DateTime StartTimeCore
             }
         }
 
+        /// <summary>Gets arguments.</summary>
+        /// <param name="startInfo">The start info with which to start the process.</param>
+        private string[] GetArgs(ProcessStartInfo startInfo)
+        {
+            if (!startInfo.UseShellExecute)
+            {
+                return ParseArgv(startInfo);
+            }
+
+            string shellArgs = string.IsNullOrEmpty(startInfo.Arguments) ? startInfo.FileName : startInfo.FileName + " " + startInfo.Arguments;
+            return new string[2] { GetExecPath(), shellArgs };
+        }
+
+        /// <summary>Gets execution path</summary>
+        private string GetExecPath()
+        {
+            string[] allowedProgramsToRun = { "xdg-open", "gnome-open", "kfmclient" };
+            foreach (var program in allowedProgramsToRun)
+            {
+                string pathToProgram = GetPathToProgram(program);
+                if (!string.IsNullOrEmpty(pathToProgram))
+                {
+                    return Path.Combine(pathToProgram, program);
+                }
+            }
+            return string.Empty;
+        }
+
+        /// <summary>
+        /// Gets the path to the program
+        /// </summary>
+        /// <param name="program"></param>
+        /// <returns></returns>
+        private string GetPathToProgram(string program)
+        {
+            string path = Environment.GetEnvironmentVariable("PATH");
+            string[] dirs = path.Split(":");
```

Comment:
If for some reason PATH is not set in this shell, this will throw NRE.

Reply:
it looks like Mono falls back to the current directory.
https://github.com/mono/mono/blob/master/mono/eglib/gpath.c#L236

Reply:
You could eliminate all this code and just call `ResolvePath()` if you don't mind a slightly different resolution algorithm. I think that's fine.

Reply:
yes, I just noticed ResolvePath() is also calling path to get path. I'm going to extract that logic into a private method and reuse that private method both in ResolvePath() and in here. Thanks